### ISSUE TITLE
Add API to switch focus from `sidePanel` to main document to enable accessibility features

### ISSUE THREAD

**Issue by SpaceK33z** [hober/web-extensions#693]
**Description:**
Interacting with the extension’s side panel (`chrome.sidePanel`) shifts focus away from the main webpage, and there is no way to programmatically return focus to the main document. This limitation makes it impossible to implement some features related to accessibility.

**Context:**
Managing focus between the main document and the side panel's document within the same tab is essential for accessibility. 
With proper focus management, the user can select something from the side panel, and the side panel can simulate certain events that only work when the main document is in focus.

Currently, extensions can switch focus between tabs, but not within the main document and side panel document.

**Proposed Solution:**
Introduce a new API that allows extensions to programmatically shift focus from the side panel back to the main document.

**Comment by hanguokai** [comment]
In contrast, when the side panel is opening, the focus remains on main webpage and does not move to the side panel, which makes some functions impossible. For example, a typical user requirement is to focus on the input box when the side panel opens, so that the user can input something immediately.

So I hope there is a way to solve the problem of bidirectional focus movement.

**Comment by oliverdunk** [comment]
We discussed this during an in-person triage meeting at TPAC. We are aligned on the following:

- The behavior @hanguokai mentioned seems like a bug. We should look at ensuring the side panel takes focus from the page when it is first opened regardless of the mechanism used to open it.
- We are supportive of implementing a tabs.focus() method which would return focus to the web content. This would close the popup if open and move focus away from other parts of the UI such as the address bar.

**Comment by hanguokai** [comment]
> We are supportive of implementing a tabs.focus() method which would return focus to the web content. This would close the popup if open and move focus away from other parts of the UI such as the address bar.

A typical case comes to mind. When opening a new tab page(by shortcut or mouse), the focus is on the address bar for user input. Now the new tab page extension would like to call tabs.focus() to move the focus into the NTP for its own searching or shortcut function. Do browsers allow this behavior?

There may be other risks of abuse.

**Comment by Rob--W** [comment]
I raised that exact question and the answer is that the content area would be focused indeed. The meeting notes for the TPAC 2024 meeting will be published soon, here is the relevant excerpt:

* [rob] If the address bar was focused and you call this method, what happens?
* [devlin] It focuses the main content in the tab (document tree in Chrome, webview in Safari) - whatever was the active frame is focused.

**Comment by jpc-ae** [comment]
Any movement on this?

**Comment by Diego1149** [comment]
+1, this is blocking one of our launches.

**Comment by qupig** [comment]
> We are supportive of implementing a `tabs.focus()` method which would return focus to the web content. This would close the popup if open and move focus away from other parts of the UI such as the address bar.

`tabs.focus()` would be nice, but I think there is a need for a capability to enable the extension to switch focus arbitrarily.

Here is an example, hoping that the focus will remain in `sidebar`:

A common use of `sidebar` is to display a list of vertical tabs, and clicking on one of those tabs invokes `browser.tabs.update(tabId, { active: true });` to activate the corresponding `tab`.

However, currently after calling `tabs.update`, the focus will auto switch to the `tab`, causing `sidebar` to lose focus.

If I want to use the keyboard up and down keys in the `sidebar` to switch between different tabs, this is currently NOT possible. 

I'm sure there are plenty of other situations there that would like to keep the focus in the `sidebar` after switching tab.

So we need a method to prevent the `tabs.update` call from shifting focus, or we also need a `sidebar.focus()` method as well.


> Now the new tab page extension would like to call tabs.focus() to move the focus into the NTP for its own searching or shortcut function. Do browsers allow this behavior?
> There may be other risks of abuse.

I support the extension having that ability.

The `newtab` page implemented by the extension can implement its own omnibox, and may be richer or have different features than the browser's own.

If this is the user expects to replace the default address bar focus, an extra click will disrupt the user workflow.

Preventing abuse is something extensions should provide relevant options for.
Otherwise the user can refuse to use the extension.

**Comment by remusris** [comment]
I'm having this exact same issue and yes it would be optimal to change the default so that the sidepanel gets into focus upon trigger or at least an option to focus the sidepanel when opening it.  There's only two workarounds I'm aware of, on Windows/Linux you can press `Shift + F6` to change focus to the sidepanel (`F6` is the reverse direction) while on mac is `Command + Option + Arrow Up / Arrow Down` but in most cases it seems that `Command + Option + Arrow Up` will get you there without needing to highlight the address bar.  The other alternative is just to use the popup window for a single function if you really need to have autofocus working and trigger `chrome.actions.openPopup()` via some keyboard shortcut from a chrome command.  

Below is also a link pertinent to this dialogue
https://groups.google.com/a/chromium.org/g/chromium-extensions/c/nb058-YrrWc

**Comment by qupig** [comment]
@remusris Yes, I know those shortcuts could switch tabs, but that's only in the order of the tab bar.

In the extension, the sidebar can list a list of completely different orders from the tab bar, and losing focus can be a disaster in the user experience.

Based on specific use cases, some operations will not be very good when putting in popup, so I think we still need this ability.

UPDATE:

Sorry, I thought you were talking about `Command + Option + Left/Right`.

However, just tried and `Command + Option + Arrow Up / Arrow Down` does not work with me.

Even if it works, it is not advisable to require additional operations every time.

**Comment by remusris** [comment]
@qupig 
I agree, the workarounds I listed are imperfect and not optimal.  I'm building an app right now to where this functionality seriously hampers it and decreases the UX substantially. I'm not too sure how this can be escalated to someone on the chromium team but if you have any ideas feel free to reach out to me (check personal website in profile).

**Comment by oliverdunk** [comment]
Hi all, I just wanted to confirm that I've seen your feedback. We haven't looked at making any changes yet, but I'll make sure this stays on the team's radar and that the other requests (e.g focusing a tab without switching focus) is part of any discussions. Thanks for your patience and sorry this hasn't moved anywhere yet.

**Comment by brianshultz** [comment]
@oliverdunk Any update on the Chrome side of things? Know it's on the team's radar and appreciate all the browsers being supportive, but wanted to add a friendly nudge here since the workarounds are pretty limited outside of using the debugger API 🫤

**Comment by SpaceK33z** [comment]
@dotproto published a PR with a proposal here https://github.com/w3c/webextensions/pull/817 — thank you!
